lib/common/cmdline.c

```diff
@@ -144,6 +144,74 @@ pcmk__add_arg_group(GOptionContext *context, const char *name,
     // cppcheck-suppress memleak
 }
 
+static gchar *
+string_replace(gchar *str, const gchar *sub, const gchar *repl)
+{
+    /* This function just replaces all occurrences of a substring
+     * with some other string.  It doesn't handle cases like overlapping,
+     * so don't get clever with it.
+     *
+     * FIXME: When glib >= 2.68 is supported, we can get rid of this
+     * function and use g_string_replace instead.
+     */
+    gchar **split = g_strsplit(str, sub, 0);
+    gchar *retval = g_strjoinv(repl, split);
+
+    g_strfreev(split);
+    return retval;
+}
+
+gchar *
+pcmk__quote_cmdline(gchar **argv)
+{
+    GString *gs = NULL;
+    gchar *retval = NULL;
+
+    if (argv == NULL || argv[0] == NULL) {
+        return NULL;
+    }
+
+    gs = g_string_sized_new(100);
+
+    /* The first item is the program name. */
+    g_string_append(gs, argv[0]);
+
+    for (int i = 1; argv[i] != NULL; i++) {
+        if (strchr(argv[i], ' ') != NULL) {
+            if (strchr(argv[i], '\'') != NULL) {
+                /* If a single quote occurs in the command line arg, replace it with
+                 * an escaped version.  We do this instead of counting on XML to handle
+                 * the escaping for two reasons:
+                 *
+                 * (1) This keeps the string as valid shell.
+                 * (2) The string we are feeding to libxml is something like:  "a b 'c d' e".
+                 *     It won't escape the single quotes around 'c d' here because there is
+                 *     no need to escape quotes inside a different form of quote.  If we
+                 *     change the string to "a b 'c'd' e", we haven't changed anything - it's
+                 *     still single quotes inside double quotes.
+                 *
+                 *     On the other hand, if we replace the single quote with "&apos;", then
+                 *     we have introduced an ampersand which libxml will escape.  This leaves
+                 *     us with "&amp;apos;" which is not what we want.
```

Comment:
Not to mention we wouldn't want to use entities except with XML/HTML output

Reply:
Updated the comment.